### DISCUSSION TITLE
printing already colorized text?

### DISCUSSION THREAD

**Discussion by jsholmes** [ArthurSonzogni/FTXUI#436] [Q&A]
Hey there.  I'm getting some text back from a library function call that is already colorized using ANSI escape sequences (I think?), but when I try to show it in my ftxui-based program, the escapes aren't escaped anymore and are instead printed. e.g.:

`at time[1m[36m14:40:34[0m[0m`

Do you have any suggestions on how I can get these to render with the colors intended?

**Comment by jsholmes** [comment]
I wonder if maybe I need to just translate all the `\033` to `\u001b` ?

**Comment by jsholmes** [comment]
Oh nope.  I see that `Utf8ToGlyphs` ignores control characters, but doesn't eat the fully escaped thing, so I'm left displaying `[1m` (bold) `[36m` (cyan) etc.

**Reply by jsholmes** [reply]
Ok.  I'm removing the ansi escape sequences with this as a stopgap measure:

```
    std::regex re{"\x1B" // ESC
                  "(?:"  // 7-bit C1 Fe (except CSI)
                  "[@-Z\\-_]"
                  "|" // or [ for CSI, followed by a control sequence
                  "\["
                  "[0-?]*" // Parameter bytes
                  "[ -/]*" // Intermediate bytes
                  "[@-~]"  // Final byte
                  ")"};
    string unescaped = std::regex_replace(story, re, "");
```
But Utf8ToGliphs is still eating the `\n`.  Is there something I should be using other than `text` or `paragraph` to render this?  Or does ftxui not have any dom element appropriate for pre-formatted text currently?  I don't really grok the internals of ftxui rendering enough to add this, but perhaps you can point me in the right direction?

As a stopgap, I'll just split the string and reassemble it into a vbox.  Maybe the right thing to do to fix all of this is parse the ansi escape sequences and return some dom element that tries to render it correctly using ftxui native stuff like `text`, `paragraph`, `color`, etc.?

**Reply by ArthurSonzogni** [reply]
Utf8ToGlyphs doesn't deals with terminal escape sequence. It split a UTF8 char string into several chunks representing a glyph to be displayed. Several char form a codepoint. Several codepoint can represent a glyph.
So, don't use Utf8ToGlyphs.

FTXUI is only a terminal client. Not a terminal server. So I haven't implemented a way to "interpret" what is sent/received to/from a terminal client. I only implemented "interpreting" what is sent/received to/from terminal.

Here is what FTXUI implement for receiving bytes from the terminal:
https://github.com/ArthurSonzogni/FTXUI/blob/master/src/ftxui/component/terminal_input_parser.cpp

I guess you need to do the same, but for receiving bytes from the terminal client instead.

Your regexp is likely a good solution, as long as it is scoped to your particular library.
